device-api/src/topology/mod.rs

```diff
@@ -0,0 +1,349 @@
+    // Note that this does not guarantee devices are attached to the same PCIE switch.
+    // More switches could exist under the host bridge switch.
+    LinkTypeHostBridge = 30,
+
+    // NOTE(@bg): Score 40 and 50 is reserved for LinkTypeMultiSwitch and LinkTypeSingleSwitch.
+    // NOTE(@bg): Score 60 is reserved for LinkTypeBoard
+
+    // LinkTypeSoc two devices are on the same Soc chip.
+    LinkTypeSoc = 70,
+}
+
+impl LinkType {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            LinkTypeUnknown => "LinkTypeUnknown",
+            LinkTypeInterconnect => "LinkTypeInterconnect",
+            LinkTypeCPU => "LinkTypeCPU",
+            LinkTypeHostBridge => "LinkTypeHostBridge",
+            LinkTypeSoc => "LinkTypeSoc",
+        }
+    }
+}
+
+pub trait HardwareTopologyHint {
+    fn get_hw_topology_hint(&self) -> String;
+}
+
+pub struct Topology {
+    hwloc_topology: Box<dyn Hwloc>,
+    topology_matrix: BTreeMap<(String, String), LinkType>,
+}
+
+impl Topology {
+    pub fn new(devices: Vec<Device>) -> DeviceResult<Topology> {
+        let mut new_topology = Self {
+            hwloc_topology: Box::new(HwlocTopology::new()),
+            topology_matrix: BTreeMap::new(),
+        };
+
+        let keys = devices.iter().map(|d| d.busname().unwrap()).collect();
+        new_topology.populate_with_keys(keys)?;
+
+        Ok(new_topology)
+    }
+
+    fn populate_with_keys(&mut self, devices: Vec<String>) -> DeviceResult<()> {
+        // Initialize hwloc topology
+        self.hwloc_topology.init_topology()?;
+
+        // Set I/O types filter
+        self.hwloc_topology
+            .set_io_types_filter(hwloc_type_filter_e_HWLOC_TYPE_FILTER_KEEP_IMPORTANT)?;
+
+        // Load the topology
+        self.hwloc_topology.load_topology()?;
+
+        match self.populate_topology_matrix(devices) {
+            Ok(matrix) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn populate_topology_matrix(&mut self, devices: Vec<String>) -> DeviceResult<()> {
+        for i in 0..devices.len() {
+            for j in 0..devices.len() {
+                let dev1_bdf = devices.get(i).unwrap().clone();
+                let dev2_bdf = devices.get(j).unwrap().clone();
+
+                let link_type = self.search_interconnect(&dev1_bdf, &dev2_bdf);
+                let key = if dev1_bdf > dev2_bdf {
+                    (dev2_bdf, dev1_bdf)
+                } else {
+                    (dev1_bdf, dev2_bdf)
+                };
+
+                self.topology_matrix.insert(key, link_type);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn search_interconnect(&self, dev1_bdf: &str, dev2_bdf: &str) -> LinkType {
+        unsafe {
+            if dev1_bdf == dev2_bdf {
+                return LinkTypeSoc;
+            }
+
+            let ancestor_obj = self
+                .hwloc_topology
+                .get_common_ancestor_obj(dev1_bdf, dev2_bdf)
+                .unwrap();
```

Comment:
Change the return type of `search_interconnect` to `DeviceResult`.
```suggestion
                .get_common_ancestor_obj(dev1_bdf, dev2_bdf)?;
```

Reply:
applied in eb2fd550a51a815e6c6eeebfc6b20649754a1a70